app/policies/mhv_messaging_policy.rb

```diff
@@ -1,9 +1,26 @@
 # frozen_string_literal: true
 
 MHVMessagingPolicy = Struct.new(:user, :mhv_messaging) do
-  SM_ACCOUNT_TYPES = %w[Premium].freeze
-
   def access?
-    SM_ACCOUNT_TYPES.include?(user.mhv_account_type) && user.va_patient?
+    client = SM::Client.new(session: { user_id: user.mhv_correlation_id })
+    validate_client(client)
+  end
+
+  def mobile_access?
+    client = Mobile::V0::Messaging::Client.new(session: { user_id: user.mhv_correlation_id })
```

Comment:
Curious if there were any concerns about performance by calling an external system potentially with every authorized route.

The usage on mobile is orders of magnitude greater than any usage on MHV today, so if this is called as part of user accessible services call, the service is probably going to see significantly more traffic than it ever has seen before

Reply:
@aherzberg I think from a code perspective I think it looks okay actually after removing the param from the policy function. So I think from code quality standpoint it's fine now

Reply:
@bosawt there are definitely performance concerns, especially since one of these methods is called after login and has the ability to block the user from using the app. andrew and i were just talking about that yesterday. Here's the authenticate method:
```
        LOCK_RETRY_DELAY = 0.3 # Number of seconds to wait between attempts to acquire a session lock
        RETRY_ATTEMPTS = 40 # How many times to attempt await of acquiring a session lock by a preceding request
        
        def authenticate
          raise 'ICN is required for session creation' unless user_key

          iteration = 0

          # Loop unless a complete, valid MHV session exists, or until max_iterations is reached
          while invalid?(session) && iteration < RETRY_ATTEMPTS
            break if lock_and_get_session # Break out of the loop once a new session is created.

            sleep(LOCK_RETRY_DELAY)

            # Refresh the MHV session reference in case another thread has updated it.
            refresh_session(session)
            iteration += 1
          end
          if invalid?(session) && iteration >= RETRY_ATTEMPTS
            Rails.logger.info("Failed to create #{@client_session} after #{iteration} attempts to acquire lock")
          end

          self
        end
```
So if MHV is down, it will spend over 10 seconds just sleeping while it makes the 40 requests, plus all the time needed for making the actual requests. sleeping blocks a thread, so if mhv is down, i'd imagine it will absolutely maul the servers. but afaik we kinda have to use this. Maybe it would be worth making an alternative version of the authenticate method that only tries once. I'm definitely open to ideas.

Reply:
yeah, that could be an issue. I'm not familiar with the volume MHV currently handles vs Mobile login volume but we could look into a few thing to mitigate this. For login authorization only, we could reduce the number of retries allowed and then if it fails on all those retries, fall back to the legacy policy that doesn't have an upstream call. We also have this behind a flipper so we can find where that MHV capacity is at and then keep the % of users low enough to not bring down the service.

Reply:
@aherzberg i definitely think we need to find a way of ensuring that this can't block the user accessible services endpoint because that would block the mobile app entirely. because the user accessible services endpoint isn't turned off by maintenance windows, so if there's an mhv maintenance window, the mobile app will continue to try to fetch the services list, get timeouts, and block use of the mobile app. it's less of an issue for the messaging endpoints, but i think it's really bad for fetching that accessible services list

Reply:
I've added some logic to set the user accessible services endpoint to only try 2 times and if it fails it will fallback to the legacy policy. Do you think that's enough to address the issue?

Reply:
I would not fall back to the legacy policy. Earlier, you said the legacy policy "doesn't have an upstream call". That's not true. The policy relies on this user method:
```
  def mhv_account_type
    identity.mhv_account_type || MHVAccountTypeService.new(self).mhv_account_type
  end
```
It's hard for me to be certain, but I think it's possible that `identity.mhv_account_type` is always nil. I don't see it ever being set. Regardless, it's clear from that code that we can't rely on it being set. If you look at `MHVAccountTypeService`, you'll eventually see that the value comes from this snippet, which is where the bug that's prompting this ticket lies. Specifically, the bug is that we receive nil eligible_data_classes back.
```
      bb_client = BB::Client.new(session: { user_id: @user.mhv_correlation_id })
      bb_client.authenticate
      bb_client.get_eligible_data_classes.members.map(&:name)
```
If you look at that BB Client, you'll see that it includes the `Common::Client::Concerns::MHVSessionBasedClient`, which is the same class the mhv policy uses, and that `authenticate` method in the code snippet above is the same one we're using to validate that the user can use mhv. And that method does make an upstream call. The big differences between authenticating with bb vs what you have in this PR are that 1) legacy doesn't use the mobile client config and as a result uses a different cache 2) the legacy approach calls `get_eligible_data_classes` after authenticating, which is where the bug lies. So it makes the same upstream call but yields worse results. 

As to how many times to retry... I lean toward not retrying. The legacy policy uses the same method and doesn't retry. I'd maybe add some temporary logging to the authenticate method to get more insight into how often requests fail. Currently, the logging is only for when it fails 40 times, which btw has happened 8 times in the past day.

Reply:
It occurred to me last night that since the legacy policy makes the same upstream request, we should already be experiencing issues if the 40 retries are problematic. I checked DD this morning. Interestingly, the authorized services endpoint has a 0% error rate over the past month and a max latency of 5.74s. Average latency is more like 26ms and p95 is 1.46s. By vets-api standards, those numbers aren't bad. If you look at the latency breakdown, there's some pretty useful information. First, I see that roughly half of requests don't make upstream calls at all,  which means that in this user method I pointed out yesterday:
```
  def mhv_account_type
    identity.mhv_account_type || MHVAccountTypeService.new(self).mhv_account_type
  end
```
either `identity.mhv_account_type` is set or the value is cached (MHVAccountTypeService uses a cache). Second, the upstream mhv session call does vary in length, but I haven't found any instances of it failing outright and the longest latency on the session request was about 1s. But the legacy policy also has to request eligible classes after establishing the session. That call frequently fails, usually taking about 200ms per failure, and when it fails it seems to retry twice. So a lot of the latency actually comes from that request. 

So this alleviates some of my concerns about the retries. I don't understand why it never fails completely, especially given that I was under the impression that MHV has maintenance windows every weekend that I would expect would cause failures. In light of this information, I think we should not change the retry logic at all just let it happen because it's probably better to have some occasional latency due to retries than to reduce the user's access. 

I also think we should actively monitor this endpoint on datadog for a while and note two things:
1) does the percentage of requests that have to make upstream calls change? It may increase because we'll be using a different cache that may not already be set. IMO, the way caching works here is kinda silly. We're establishing this session in multiple places with different caches, which seems inefficient, but that's beside the point here.
2) does the overall latency change? if we use the cache less frequently per point 1, the p50 may increase. but we'll also be removing the eligible classes calls, which are responsible for a lot of the latency. I think we may experience an overall latency decrease.

Reply:
That all makes sense. all good. makes sense to me. Yeah im not sure exactly why the upstream calls don't cause failures either. Maybe errors get a caught and the timeout time for the user endpoint is long enough to finish even after all those retries? Luckily with the flipper we can incrementally increase usage to make sure we don't bring down the endpoint completely. I'll go ahead and revert the last commit.